### PR TITLE
enable broker bootstrap methods to be provided by dso plugins, and drop compiled-in pmix support

### PR DESCRIPTION
**Pull request by garlick**
Problem: sites that require pmix to bootstrap flux must remember to build flux-core with pmix support, but pmix support is experimental and may be more effectively provided by an external plugin.

Since pmix server support is already implemented as a shell plugin in the [flux-pmix](https://github.com/flux-framework/flux-pmix) project, drop the pmix client support from flux core (and the `--enable-pmix-bootstrap` option) and change the PMI client library to allow PMI methods to be provided by dso plugins.  A follow-on PR to flux-pmix will add such a plugin.

This PR also adds support for a `broker.boot-method` attribute which may be used to force the broker to use a particular method like
```
jsrun ... flux start -o,-Sbroker.boot-method=pmix ...
```

Alternatively, a site that wants Flux to try to bootstrap from pmix when simple PMI is not available may set this in the environment:
```
FLUX_PMI_CLIENT_METHODS="simple pmix single"
```
Edit: the other methods remain there so that the foreign-launched flux can launch sub-instances with native methods.

### CODE REVIEW

**Comment by grondo** [comment]
> Edit: the other methods remain there so that the foreign-launched flux can launch sub-instances with native methods.

If the `simple` method should always be first, and `single` last, should something like `FLUX_PMI_METHODS_EXTRA="pmix"` be supported instead, to make it less likely sites would forget the correct order or drop one of the important methods? I haven't fully digested the PR yet though, so maybe there are some cases where it is valid to remove the simple and single methods.

**Comment by garlick** [comment]
Good thought.  I think testing is the main case for letting the entire order be configurable.  Hmm, requires more thought if there's a safer way to handle the testing...

**Comment by grondo** [comment]
Seems fine if necessary for testing. I don't have a good feeling for how often this would be used, or if my fear of getting the order wrong is even founded. Seems fine to leave it as is for now and address any issues if they come up.

**Comment by garlick** [comment]
I should mention @ggouaillardet here as he added the initial pmix bootstrap support and has a use case.  Perhaps he'll have some feedback on the proposed change?  Another place where this could potentially be used is on our sierra system, but we can coach the admins on that system to ensure the environment settings are correct.

**Review by grondo — APPROVED** [review]
This was a large PR, but I looked through it and nothing much jumped out. Seems like a huge improvement that all the PMIx stuff is now located in the flux-pmix repo.

Just a couple initial thoughts inline. But otherwise, seems fine to merge this one. I did try this branch out on fluke and corona, launching both under the old version of flux, and the version from this branch. I didn't have a chance to try launching under Slurm, but I do not expect any issues.

File: src/broker/doc/bootstrap.md [hunk 1]
```diff
@@ -68,13 +68,12 @@ trace on stderr from the start command.
...
-2. if configured with `--enable-pmix-bootstrap`, and `PMIX_SERVER_URI`
-or `PMIX_SERVER_URI2` is set, use PMIx
-3. if `libpmi.so` can be dynamically loaded, use PMI-1 (a specific library
-name/path may be forced by setting `PMI_LIBRARY`)
+The code in `boot_pmi.c` (using libpmi/upmi.c) attempts to adapt the broker's
+PMI client to different situations if the PMI-1 wire protocol is not available.
+It tries the following in order:
+1. simple PMI wire protocol
+2. find a PMI-1 library
+3. pmix (if a suitable plugin is found)
```
[the remark this answers is not in the document]
Comment (garlick) [reply src/broker/doc/bootstrap.md:76]: Good catch!  Will fix.

File: src/broker/runat.c [hunk 2]
```diff
@@ -75,7 +75,6 @@ static const char *env_blocklist[] = {
     "PMI_FD",
     "PMI_RANK",
     "PMI_SIZE",
-#ifdef HAVE_LIBPMIX
```
[the remark this answers is not in the document]
Comment (garlick) [reply src/broker/runat.c:78]: Hmm that makes sense, although I'm a little unclear on the best way to arrange that.  I'll open an issue.

**Comment by garlick** [comment]
whoops need to do that again - I pushed an older working copy

File: src/broker/runat.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (grondo) [reply src/broker/runat.c:78]: Is the plugin running in the address space of the broker? In that case the plugin could just take care of unsetting any PMIX environment variables once it is done with them? Of course, if these variables should be unset even if PMIx is not being used then it makes sense to keep these here, so maybe nothing to do at this point..

[the remark this answers is not in the document]
Comment (garlick) [reply src/broker/runat.c:78]: Oh great idea.  I hadn't thought of that.

I'll go ahead and drop the PMIX_ removals from this PR and I can work through this over in flux-pmix.

**Comment by garlick** [comment]
Thanks - I'm going to set MWP on this one.

**Pull request merged**